### ISSUE TITLE
FakeValuesService is not thread-safe

### ISSUE THREAD

**Issue by rwong-gw** [datafaker-net/datafaker#759]
**Describe the bug**
Using Gatling to simulate hundreds of users making requests to a server, data will occasionally become corrupted, causing all threads to eventually hang or deadlock. A thread dump taken after a deadlock indicates a problem with `java.util.WeakHashMap`, which is known to be [not thread-safe](http://dowbecki.com/WeakHashMap-is-not-thread-safe/):
```
"Thread-62" prio=5 Id=78 RUNNABLE
	at java.base@11.0.15/java.util.WeakHashMap.transfer(WeakHashMap.java:529)
	at java.base@11.0.15/java.util.WeakHashMap.resize(WeakHashMap.java:506)
	at java.base@11.0.15/java.util.WeakHashMap.put(WeakHashMap.java:467)
	at app//net.datafaker.service.FakeValuesService.split(FakeValuesService.java:246)
	at app//net.datafaker.service.FakeValuesService.fetchObject(FakeValuesService.java:195)
	at app//net.datafaker.service.FakeValuesService.safeFetch(FakeValuesService.java:155)
	at app//net.datafaker.service.FakeValuesService.resolve(FakeValuesService.java:408)
	at app//net.datafaker.providers.base.AbstractProvider.resolve(AbstractProvider.java:21)
	...

"Thread-229" prio=5 Id=245 RUNNABLE
	at java.base@11.0.15/java.util.WeakHashMap.transfer(WeakHashMap.java:529)
	at java.base@11.0.15/java.util.WeakHashMap.resize(WeakHashMap.java:506)
	at java.base@11.0.15/java.util.WeakHashMap.put(WeakHashMap.java:467)
	at app//net.datafaker.service.FakeValuesService.split(FakeValuesService.java:246)
	at app//net.datafaker.service.FakeValuesService.fetchObject(FakeValuesService.java:195)
	at app//net.datafaker.service.FakeValuesService.safeFetch(FakeValuesService.java:155)
	at app//net.datafaker.service.FakeValuesService.resolve(FakeValuesService.java:408)
	at app//net.datafaker.providers.base.AbstractProvider.resolve(AbstractProvider.java:21)
	...

"Thread-285" prio=5 Id=301 RUNNABLE
	at java.base@11.0.15/java.util.WeakHashMap.transfer(WeakHashMap.java:529)
	at java.base@11.0.15/java.util.WeakHashMap.resize(WeakHashMap.java:494)
	at java.base@11.0.15/java.util.WeakHashMap.put(WeakHashMap.java:467)
	at app//net.datafaker.service.FakeValuesService.split(FakeValuesService.java:246)
	at app//net.datafaker.service.FakeValuesService.fetchObject(FakeValuesService.java:195)
	at app//net.datafaker.service.FakeValuesService.safeFetch(FakeValuesService.java:155)
	at app//net.datafaker.service.FakeValuesService.resolve(FakeValuesService.java:408)
	at app//net.datafaker.providers.base.AbstractProvider.resolve(AbstractProvider.java:21)
	...
```
In addition to `WeakHashMap`, FakeValuesService also uses several `IdentityHashMap`s, which are also not thread-safe. Although I have not observed any deadlocks involving these, that could simply be due to the particular APIs I'm calling.

**To Reproduce**
See attached command-line program `DataFakerWeakHashMapBug.java`. The program takes two command-line arguments, the number of threads and the total number of iterations per thread. Because of the non-deterministic nature of the issue, you will likely need to run the program several times to reproduce it. On my system, I was able to get it to hang about once every ten times, using 400 threads and 10,000 iterations.

**Expected behavior**
DataFaker shouldn't cause everything to hang.

**Versions:**
 - OS: macOS Ventura 13.3
 - JDK: 11.0.15
 - Faker Version [e.g. 22]: 1.8.0

**Additional context**
[DataFakerWeakHashMapBug.java.txt](https://github.com/datafaker-net/datafaker/files/11194366/DataFakerWeakHashMapBug.java.txt)

**Comment by rwong-gw** [comment]
By the way, I have a version of FakeValuesService with my own fixes, using Guava weak-key caches to replace the WeakHashMaps and ConcurrentHashMaps to replace the IdentityHashMaps, but it's for the 1.8.0 version of DataFaker, as I don't expect my company to switch to Java 17 anytime soon. Please let me know if you would like me to post a PR.

**Comment by bodiam** [comment]
Absolutely, that would be great. 1.8 is sort of no longer maintained, but if you have a patch, I think I'll be able to create an 1.8.2 version for you.

**Comment by kingthorin** [comment]
@rwong-gw are you stuck on Java 8?

**Comment by rwong-gw** [comment]
@kingthorin No, I'm on Java 11.

Fortunately, there are only a handful of differences between the 1.x and 2.x versions of FakeValuesService, so it shouldn't be too hard to merge my changes (if they're approved) back into the main branch.

@bodiam I have opened PR #760 for this. Let me know if you have any questions.

**Comment by bodiam** [comment]
Thanks for this PR! I do have a question: any chance to do it without Guava? 

We could shadow the jar, but the idea was to have almost zero dependencies. Adding Guava conflicts a bit with that.

**Comment by rwong-gw** [comment]
I chose to use Guava because it supported weak-key caches, which should be reasonably equivalent to using WeakHashMaps for caching.

Instead of using Guava, I will post an update that just wraps the existing WeakHashMaps with Collections.synchronizedMap().

**Comment by rwong-gw** [comment]
I have updated my PR to remove the Guava dependency and wrap the WeakHashMaps with Collections.synchronizedMap(). Let me know if you have any other questions.

**Comment by snuyanzin** [comment]
Sorry I was late here...

In fact `Faker` is NOT designed to be shared across several threads.
In case there is a need for multithreading - use one `Faker` per thread

**Comment by rwong-gw** [comment]
My team has been assuming that we could use Faker from multiple threads because of the large amount of data-loading involved. After going through the DataFaker codebase, though, I see that the English (and only the English) locale chain is shared. Maybe we can live with that for now, but since my company has clients all over the world, it will limit our ability to use DataFaker. Are there any plans to make non-English locale chains shared?

Also, suggest your documentation mention (if not emphasize) that Faker cannot be shared between threads.

**Comment by snuyanzin** [comment]
>After going through the DataFaker codebase, though, I see that the English (and only the English) locale chain is shared.

Can you elaborate here, may be with some examples/use cases?

**Comment by bodiam** [comment]
Why do you want non English locales to be shared? The amount of data for non English locales is very limited if that's a concern?

**Comment by snuyanzin** [comment]
I was asking not about how it's implemented in datafaker...
I was asking about use-cases and code snippet highlighting some situations where it does behave in a wrong way

**Comment by snuyanzin** [comment]
btw:
>Are there any plans to make non-English locale chains shared?

have you considered `doWith` methods?

**Comment by snuyanzin** [comment]
are there only perf issues or functional issues as well?
In case there are functional issues can you provide a reproducible junit test?

**Comment by rwong-gw** [comment]
`doWith` ignoring and erasing `addPath` or `addUrl` additions is a functional issue.

I have never used `doWith` myself, and don't intend to, since it doesn't apply to my use case. I only noticed the problem from reading the source code, and am reporting it to you as a courtesy.

**Comment by snuyanzin** [comment]
I do not understand you
can you show where exactly you see a problem?

I created a dummy test (I added system println for simplicity) based on `net.datafaker.providers.base.CustomFakerTest#insectAntTestExpressionFromFile`
```java
    void insectAntTestExpressionFromFile() {
        MyCustomFaker myFaker = new MyCustomFaker(Locale.GERMANY);
        System.out.println("Expected German: " + myFaker.insectFromFile().ant());
        System.out.println("Expected Chinese: " + myFaker.doWith(() -> myFaker.name().firstName(), Locale.SIMPLIFIED_CHINESE));
        System.out.println("Expected German: " + myFaker.insectFromFile().ant());
        myFaker.addPath(Locale.ENGLISH, Paths.get("src/test/ants.yml"));
        System.out.println("Expected English: " + myFaker.doWith(() -> myFaker.insectFromFile().ant(), Locale.ENGLISH));
        System.out.println("Expected German: " + myFaker.insectFromFile().ant());
    }
```
and the output 
```
Expected German: Treiber Ameise
Expected Chinese: 金鑫
Expected German: Feuerameise
Expected English: Honey ant
Expected German: Treiber Ameise
```
so it still keeps info loaded from file

**Comment by rwong-gw** [comment]
My bad, I misunderstood that `updateFakeValuesInterfaceMap` only changes an entry for a locale if it hasn't already been initialized.

**Comment by snuyanzin** [comment]
ok...
now i hope you see why it's much better to provide proof like e.g. reproducible tests rather than just `code reading`

**Comment by snuyanzin** [comment]
>The doWith methods have their own problems.
>
>As I understand them, they are intended to temporarily switch to a different locale, so they wouldn't apply to the above use case. If you are using doWith to switch between multiple languages, each use of a non-English locale will allocate a brand-new FakeValues instance, which will load its own copy of the data, resulting in excessive file I/O and garbage collection.
>
>Also, the doWith methods wipe out any custom data that has been added with addPath or addUrl.

so as it was realized just above it is not true and there is no known issue.


Thus the only issue i see within this task is 
>Each user has its own Faker instance. Because of the getCachedFakeValue implementation, the Faker will allocate a new FakeValues instance, which will load its own copy of the data. Even though the data is identical and unchanging for every user, there will be up to N copies of it in memory.
 
 still questionable how much benefit the resolution will give... need to check

**Comment by rwong-gw** [comment]
I am still hoping you are open to making FakeValuesService thread-safe.

I took a look at the old JavaFaker version of this class, and I can see that because it maintains far less mutable state, it is much more thread-safe. IMHO, it looks like DataFaker has sacrificed thread safety in favor of performance benchmarks. This is a significant feature regression, and I don't know if my team would have adopted DataFaker if we had known about this.

**Comment by kingthorin** [comment]
We’d be happy to consider some quantitative metrics or test code if you’ve got it.

Do you have any stats on the io or memory concerns you’ve mentioned?

**Comment by snuyanzin** [comment]
>I took a look at the old JavaFaker version of this class, and I can see that because it maintains far less mutable state, it is much more thread-safe

not really, otherwise how do you explain e.g. this https://github.com/datafaker-net/datafaker/issues/574, if I remember correctly there are some more similar issues fixed

+1 for what @kingthorin wrote

**Comment by rwong-gw** [comment]
The attached program shows how memory usage and file input time increase linearly with the number of Faker instances with non-English locales.

To configure a locale with a non-trivial amount of data, I copied DataFaker's `en-US.yml` resource (a 1.4 MB file) and renamed the file and root node for a language `foo`.

On my M1 Mac, each iteration of the for-loop uses an additional 10 MB of memory and takes about 300 ms to load its data.

When the program is instead run with input arg `en`, each iteration only uses an additional 2.6 KB and takes about 3-5 ms to finish.

[DataFakerMemoryDemo.java.txt](https://github.com/datafaker-net/datafaker/files/11236755/DataFakerMemoryDemo.java.txt)

**Comment by rwong-gw** [comment]
@snuyanzin #574 is what motivated my team to switch to DataFaker. Unless you are able to point me to some of those other similar issues you mention, I still maintain that JavaFaker was essentially thread-safe. If it were still actively maintained and we were able to get the bug fixed over there, there would have been much less need for us to make the switch. For our use case, fake data generation is a very small portion of the work done by a thread, so there is no need for it to be highly performant.

**Comment by snuyanzin** [comment]
I do not have a goal to make you using datafaker, moreover none of maitainers here are paid.
In case you have questions/issues we can help you based on our capacity.

However please respect others: in case you make a statement - provide a proof.
In case you provide some metrics please also do some initial analysis...

>To configure a locale with a non-trivial amount of data, I copied DataFaker's en-US.yml resource (a 1.4 MB file) and renamed the file and root node for a language foo.
>
>On my M1 Mac, each iteration of the for-loop uses an additional 10 MB of memory and takes about 300 ms to load its data.
>
>When the program is instead run with input arg en, each iteration only uses an additional 2.6 KB and takes about 3-5 ms to finish.

In first case where you `copied DataFaker's en-US.yml resource (a 1.4 MB file) and renamed the file and root node for a language foo.` org.yaml` needs memory to parse this file.

in second case for input `en` instead of `foo` there will be parsed `en.yml` file which is just about 100 bytes.

What is the problem here?

just for fun I ran same test against JavaFaker and it took almost a minute with much more memory consumed.

So I have no idea what you want to say with the file you've attached

**Comment by snuyanzin** [comment]
Based on measurements with help of https://github.com/datafaker-net/datafaker-benchmark it shows that the approach with Guava caches and other synchronized maps leads to **drastic** performance overhead. 


e.g.
`Datafaker_KotlinFakerBenchmark.kotlinFakerTest`
before the change
```
Benchmark                                       Mode  Cnt    Score   Error  Units
Datafaker_KotlinFakerBenchmark.kotlinFakerTest  avgt   10  410.411 ± 6.429  ms/op
```
after the change (about 5x times **slower**)
```
Benchmark                                       Mode  Cnt     Score    Error  Units
Datafaker_KotlinFakerBenchmark.kotlinFakerTest  avgt   10  2268.064 ± 57.392  ms/op
```

`DatafakerSimpleMethods.firstname`
before the change
```
Benchmark                              Mode  Cnt      Score     Error   Units
DatafakerSimpleMethods.firstname      thrpt   10   5422.936 ±  33.680  ops/ms
```
after the change (about 8x times **slower**)
```
Benchmark                              Mode  Cnt      Score      Error   Units
DatafakerSimpleMethods.firstname      thrpt   10    631.485 ±   47.711  ops/ms
```

`DatafakerSimpleMethods.fullname`
before the change
```
Benchmark                              Mode  Cnt      Score     Error   Units
DatafakerSimpleMethods.fullname       thrpt   10   2410.325 ±  97.869  ops/ms
```
after the change (about 5x times **slower**)
```
Benchmark                              Mode  Cnt      Score      Error   Units
DatafakerSimpleMethods.fullname       thrpt   10    443.313 ±   15.612  ops/ms
```

`DatafakerSimpleMethods.streetAddress`
before the change
```
DatafakerSimpleMethods.streetAddress  thrpt   10   1611.112 ±  83.651  ops/ms
```
after the change (more than 2x times **slower**)
```
DatafakerSimpleMethods.streetAddress  thrpt   10    638.357 ±    8.067  ops/ms
```

also generic test `net.datafaker.benchmark.generate_one_hour.DataFakerGenerate100` is about 2 times **slower** after the change


such performance overhead is **UNACCEPTABLE**

**Comment by snuyanzin** [comment]
there is a much faster version https://github.com/datafaker-net/datafaker/pull/770 making it thread safe

**Comment by rwong-gw** [comment]
Thanks for digging into this (and your patience with me!).

What I was trying to show with the DataFakerMemoryDemo I posted was how new Faker instances with non-English locales bypass DataFaker's caching and always reload the locale's YAML data. For each iteration, the time and memory usage with the English locale is minuscule, while they increase significantly with the non-English locale. Perhaps I am misunderstanding how international users might use DataFaker, but if they were diligent, couldn't they add a large amount of data tailored for their own region/language, similar to the large en-US.yml file? Or what if someone translates the babylon5.yml quotations (500k of data) to their own language, or has a huge archive of Simpsons quotes _en Español_?

To sum up:
* The best way to use DataFaker is simply to create a new `Faker` instance and generate your data. There's no need for complicated schemes to share instances between threads, since DataFaker caches everything for you.
* Possible feature enhancement for a future release: caching of non-English data.
* The new version of FakeValuesService using StampedLockMaps looks really nice. I will try to give it a spin in my environment.

Thanks again!

**Comment by snuyanzin** [comment]
i guess the approach of having one faker and sharing it came from javafaker where creation of faker instance is insanely heavy operation (it tries to read all the files from resources and init every provider)
In datafaker faker is lazy and creation of faker is a way much faster. So yes it's much easier to create a separate instance for each thread.

Currently i don't see the issue at least by fast looking at the code... may be it exists however not obvious... so would be better to see it as a problem e.g. in profiler's output for some  cases at least

**Comment by rwong-gw** [comment]
I see some changes from the 1.8.x version of `FakeValuesService#getCachedFakeValue` and `#addUrl` where instead of creating a brand-new `FakeValues` instance, you're calling `FakeValues.of(FakeValuesContext.of(locale.getLocale()))`, so it looks like you've already implemented caching of non-English data.

**Comment by snuyanzin** [comment]
By the way @rwong-gw can you check whether current solution #770 passes your Gatling tests?
It passes the test you've attached however I didn't check anything else

**Comment by rwong-gw** [comment]
@snuyanzin For the past couple days, I have been using CI to run my Gatling tests every half hour, and everything looks good, with no hanging or deadlocks.

To assess performance, I have also been using JProfiler to compare a handful of different implementations:

1. New FakeValuesService instance (out-of-the box, DataFaker 1.8.1) per Gatling virtual user session. Because each instance re-resolves DataFaker expressions and builds its own caches, **CPU load (10% user, 19% system)** was significantly greater than the below shared implementations, and memory usage started increasing drastically after a couple minutes, eventually resulting in an OutOfMemoryError after reaching 16 GB.
2. Shared FakeValuesService with synchronized methods. **CPU load: 2-3% user, 14% system. Memory: 0.2 GB committed**.
3. Shared FakeValuesService using ConcurrentHashMaps, Guava caches (my PR #760). **CPU load: 2% user, 14% system. Memory: 0.32 GB committed**.
4. Shared FakeValuesService using StampedLockMaps (PR #770). **CPU load: 0.23% user, 10% system. Memory: 0.23 GB committed**.

Because DataFaker calls are only a small portion of the work done in my Gatling tests, the numbers for the shared implementations (2-4 above) are rather tiny, so it's difficult to tell if the difference between them is significant. Intuitively, though, it makes sense that the synchronized one would have the worst performance and the StampedLockMap one would have the best.

I have one suggestion for improvement that I'll post on the PR.

**Comment by snuyanzin** [comment]
Thanks for checking this a detailed analysis.

great that there is no hanging or deadlocks.

**Comment by bodiam** [comment]
Can this issue be closed?

**Comment by snuyanzin** [comment]
i guess once the corresponding PR is merged it could be closed

**Comment by snuyanzin** [comment]
merging since it contains a test provided at #759 
it was passed more than 30 times in a row.
There was only one failure not related to current issue (related to github actions)


Feel free to reopen in case of issues

**Comment by snuyanzin** [comment]
also backported to 1.x as
https://github.com/datafaker-net/datafaker/commit/d9c2cc262d29deb3552c9f49ab71edf0b3f6dc30
